language/diem-tools/prefetched-transaction-replay/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+            config.build_dir,
+            config.storage_dir,
+        )?;
+
+        let rpc_client = BlockingClient::new(&config.rpc_url);
+
+        Ok(PrefetchTransactionReplayer {
+            state,
+            debugger,
+            rpc_client,
+        })
+    }
+
+    fn get_keys_read_for_tx(
+        &self,
+        tx: &Transaction,
+        modules: &[CompiledModule],
+    ) -> Result<Vec<ResourceKey>> {
+        let signed_tx = match tx {
+            Transaction::UserTransaction(u) => u,
+            _ => bail!("can only process user transactions"),
+        };
+
+        let rw_analysis = ReadWriteSetAnalysis::new(read_write_set::analyze(modules)?);
+        rw_analysis.get_keys_read(signed_tx, &self.state.view)
+    }
+
+    fn get_unique_addresses_from_keys(&self, key_reads: &[ResourceKey]) -> Vec<AccountAddress> {
+        key_reads
+            .iter()
+            .map(|k| k.address)
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .collect()
+    }
+
+    fn extract_account_state(
+        &self,
+        state_with_proof: &AccountStateWithProofView,
+    ) -> Option<AccountState> {
+        // NOTE: if this actually has a chance to fail, it might be better to change
+        // the function to return Result<Option<AccountState>, _>
+        state_with_proof.blob.as_ref().and_then(|bytes| {
+            bcs::from_bytes::<AccountStateBlob>(bytes)
+                .map_err(|e| anyhow!(e.to_string()))
+                .and_then(|account_state_blob| AccountState::try_from(&account_state_blob))
+                .ok()
+        })
+    }
+
+    fn fetch_accounts_storage(
+        &self,
+        addresses: &[AccountAddress],
+    ) -> Result<Vec<AccountAddressWithState>> {
+        let requests = addresses
+            .iter()
+            .map(|a| MethodRequest::get_account_state_with_proof(*a, None, None))
+            .collect();
+
+        let results = self
+            .rpc_client
+            .batch(requests)?
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let account_states: Vec<AccountAddressWithState> = results
+            .into_iter()
+            .zip(addresses.iter())
+            .filter_map(|(r, a)| match r.into_inner() {
+                MethodResponse::GetAccountStateWithProof(v) => self
+                    .extract_account_state(&v)
+                    .map(|state| AccountAddressWithState { address: *a, state }),
+                _ => None,
+            })
+            .collect();
+
+        Ok(account_states)
+    }
+
+    fn populate_storage_for_transactions(
+        &self,
+        txs: &[Transaction],
+        version: Option<u64>,
+    ) -> Result<()> {
+        let replay_version = match version {
+            Some(v) => v,
+            None => self.debugger.get_latest_version()?,
+        };
+
+        let modules = self
+            .debugger
+            .get_diem_framework_modules_at_version(replay_version, true)?;
```

Comment:
We should probably have a TODO here for downloading the modules by look at the code dependencies of `txs` instead of relying on `get_diem_framework_modules_at_version`. But I suppose this will only work when version is `None`.

Reply:
As long as we can fetch the module code at a particular version using the RPC API, I think it should be fine even when version is not the latest. It would be good to think about how we can avoid too many round trips when doing so though. I guess the simplest would simply be to add an RPC endpoint that returns all the modules on which a particular module depends or something like this

Reply:
>  I guess the simplest would simply be to add an RPC endpoint that returns all the modules on which a particular module depends or something like this

I think it might make sense to keep this logic in an SDK on the client side to prevent the server from doing too much computation, but on the flip side that does force the client to make more requests... Definitely an interesting question.

The main thing to express in the TODO is that we shouldn't rely on `get_diem_framework_modules_at_version` because it's not implementable in a hypothetical chain where modules are published under arbitrary addresses instead of a single address.